File: src/com/sterndu/json/JsonObject.java
package com.sterndu.json;

import java.util.*;
import java.util.function.Function;

public class JsonObject extends HashMap<String,Object> implements JsonValue {

	/**
	 *
	 */
	private static final long serialVersionUID = 5544447067100285990L;

	public JsonObject() {}

	public JsonObject(Map<String, ?> m) {
		super(m);
	}

	@Override
	public String toJson() {
		return toJson(o -> "\"" + o + "\"");
	}

	@Override
	public String toJson(Function<Object, String> function) {
		StringBuilder sb = new StringBuilder();
		sb.append("{\n");
		boolean b = true;
		for (Entry<String, Object> e: entrySet()) if (e.getValue()!=this) {
			if (b) b = false;
			else sb.append(",\n");
			sb.append("\"" + e.getKey() + "\" : ");
			sb.append(JsonUtil.serialize(e.getValue(), function));
		}
		return sb.toString().replace("\n", "\n\t") + "\n}";
	}

	@Override
	public String toString() {
		return toJson();
	}

}

File: src/com/sterndu/json/JsonObject.kt
@file:JvmName("JsonObject")
package com.sterndu.json

class JsonObject : HashMap<String?, Any?>, JsonValue {
	constructor() : super()
	constructor(m: Map<String, *>) : super(m)

	override fun toJson(): String {
		return toJson { "\"$it\"" }
	}

	override fun toJson(function: (Any?) -> String): String {
		val sb = StringBuilder()
		sb.append("{\n")
		var b = true
		for ((key, value) in entries) if (value !== this) {
			if (b) b = false else sb.append(",\n")
			sb.append("\"").append(key).append("\" : ")
			sb.append(serialize(value, function))
		}
		return sb.toString().replace("\n", "\n\t") + "\n}"
	}

	override fun toJsonValue(): JsonValue {
		return this
	}

	override fun toJsonValue(function: (Any?) -> String): JsonValue {
		return this
	}

	override fun toString(): String {
		return toJson()
	}

	companion object {
		private const val serialVersionUID = 5544447067100285990L
	}
}
